### ISSUE TITLE
tryKeyboard does not work

### ISSUE THREAD

**Issue by krazar** [mscdex/ssh2#909]
Hello,

We get the error `Timed out while waiting for handshake` when connecting on SFTP server.

By reading other issues here, I attempted the `tryKeyboard` option plus event handler without success.

It's a username/password connection.

Also, I spotted that we don't get the error  with an old version of ssh2. (0.4.15) which use an older version of ssh2-stream (0.0.22). But so far, we don't understand the issue.

Thank you for your help.
Regards,

Here is the code:

```
var Client = require('../lib/client').Client;

var config = {
  host:"host",
  port:"22",
  username:"username",
  password: "pwd",
  path:"path",
  tryKeyboard: true,
  debug: console.log,
};

var conn = new Client();
conn.on('ready', function() {
  console.log('Client :: ready');
  conn.sftp(function(err, sftp) {
    if (err) throw err;
    sftp.readdir('/', function(err, list) {
      if (err) throw err;
      console.dir(list);
      conn.end();
    });
  });
})
  .on('keyboard-interactive', (name, instructions, instructionsLang, prompts, finish) => {
    console.log('name %s, instructions %s, instructionsLang %s, prompts', name, instructions, instructionsLang, prompts);
    finish([config.password]);
  })
  .connect(config);

```

Here is the debug:

```
DEBUG: Local ident: 'SSH-2.0-ssh2js0.1.20'
DEBUG: Client: Trying data.blueberrywave.com on port 22 ...
DEBUG: Client: Connected
DEBUG: Parser: IN_INIT
DEBUG: Parser: IN_GREETING
DEBUG: Parser: IN_HEADER
DEBUG: Remote ident: 'SSH-2.0-GoAnywhere5.5.7'
DEBUG: Outgoing: Writing KEXINIT
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: pktLen:292,padLen:4,remainLen:288
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXINIT
DEBUG: Comparing KEXINITs ...
DEBUG: (local) KEX algorithms: ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1
DEBUG: (remote) KEX algorithms: diffie-hellman-group-exchange-sha256
DEBUG: KEX algorithm: diffie-hellman-group-exchange-sha256
DEBUG: (local) Host key formats: ssh-rsa,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521
DEBUG: (remote) Host key formats: ssh-dss,ssh-rsa
DEBUG: Host key format: ssh-rsa
DEBUG: (local) Client->Server ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com
DEBUG: (remote) Client->Server ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-cbc,aes192-cbc,aes256-cbc
DEBUG: Client->Server Cipher: aes128-ctr
DEBUG: (local) Server->Client ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com
DEBUG: (remote) Server->Client ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-cbc,aes192-cbc,aes256-cbc
DEBUG: Server->Client Cipher: aes128-ctr
DEBUG: (local) Client->Server HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG: (remote) Client->Server HMAC algorithms: hmac-sha2-256
DEBUG: Client->Server HMAC algorithm: hmac-sha2-256
DEBUG: (local) Server->Client HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG: (remote) Server->Client HMAC algorithms: hmac-sha2-256
DEBUG: Server->Client HMAC algorithm: hmac-sha2-256
DEBUG: (local) Client->Server compression algorithms: none,zlib@openssh.com,zlib
DEBUG: (remote) Client->Server compression algorithms: none,zlib
DEBUG: Client->Server compression algorithm: none
DEBUG: (local) Server->Client compression algorithms: none,zlib@openssh.com,zlib
DEBUG: (remote) Server->Client compression algorithms: none,zlib
DEBUG: Server->Client compression algorithm: none
DEBUG: Outgoing: Writing KEXDH_GEX_REQUEST
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: pktLen:1044,padLen:8,remainLen:1040
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXDH_GEX_GROUP
```
It timeout here. Then finish with:

```
DEBUG: Outgoing: Writing KEXDH_GEX_INIT
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
Error: Timed out while waiting for handshake
    at Timeout._onTimeout (/Users/krazar/dev/ssh2/lib/client.js:695:19)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
Emitted 'error' event on Client instance at:
    at Timeout._onTimeout (/Users/krazar/dev/ssh2/lib/client.js:697:14)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7) {
  level: 'client-timeout'
}

```

**Comment by mscdex** [comment]
Some things to note:

* You're using a much older `ssh2`/`ssh2-streams`
* With recent OpenSSL versions, there are now extra, more CPU-intensive checks performed when doing Diffie-Hellman key exchanges
* Modern versions of `ssh2`/`ssh2-streams` support curve25519 with node v14 (the algorithm is automatically advertised and promoted in that scenario) and i _believe_ curve25519 does not incur the previously mentioned CPU-intensive checks
* You can explicitly set `readyTimeout` in the connection config, by default it's set to `20000` (20 seconds). While I've noticed the delayed connections with the newer OpenSSL, I've never experienced the timeout issues that others have.

**Comment by krazar** [comment]
I will clarify because it was maybe not clear:
* I use the latest version available of ssh2 `0.8.9`with Node 12 (I tested Node14 since your message without success). What I meant was that it works with an older version.
* As it works with the older version of the library (max `0.4.15`) and the connection is quite fast (about 1 second ), I didn't think it was related to CPU-intensive-checks 


Anyway,  I just made a test with a readyTimeout to 40000 and it worked. 

Thank you for your help.
